### ISSUE TITLE
403 error unable to access / write to repository

### ISSUE THREAD

**Issue by shawnesquivel** [python-semantic-release/python-semantic-release#549]
## The problem

I have a private repository that I am trying to add python-semantic-release to through GitHub Actions.

When trying to increment the version number, I am getting the error that I don't have write access to the repository. I have used the Github Classic PAT and Fine grained PAT and both don't work. I gave **all** repo permissions just to ensure that I wasn't messing anything up.

I also echoed my my secrets.TOKEN and secrets.USERNAME in GitHub Actions to make sure that they were being called properly as well. 

**Why am I getting a write error even though I have generated the access token with ALL permissions to to the repository?**

<details>
  <summary>main.yml</summary>

```
name: Semantic Release

on:
  push:
    branches:
      - main

jobs:
  release:
    runs-on: ubuntu-latest
    concurrency: release

    steps:
      - uses: actions/checkout@v2
        with:
          fetch-depth: 0
      - name: Step 1 - Echo out a GitHub Actions Secret to the logs
        run: |
          echo "The GitHub Action Secret will be masked:  "
          echo ${{ secrets.TOKEN }}
          echo "Trick to echo GitHub Actions Secret:  "
          echo ${{secrets.TOKEN}} | sed 's/./& /g'
          echo ${{secrets.USERNAME}} | sed 's/./& /g'
      - name: Print repository URL
        run: |
          echo $(git remote get-url origin)
      - name: Python Semantic Release
        uses: relekang/python-semantic-release@master
        with:
          github_token: ${{ secrets.TOKEN }}
          repository_username: __token__
```
</details>

<details>
  <summary>pyproject.toml</summary>

```
[tool.semantic_release]
version_variable = "setup.py:__version__"
branch = "main"
upload_to_repository = false

```
</details>

<details>
  <summary>setup.py</summary>

```
from setuptools import setup

__version__ = "1.0.1"

setup(
   name="pmp-otk",
   version=__version__,
   # And so on...!!!!!!
)

```
</details>

<details>
  <summary>debug log</summary>

```
debug: * We fixed the damn bug ([`6d6667a`](https://github.com/***/pmp-otk-sandbox/commit/6d6667afde48fbd3cbdabaa048989379b7216ea9))')
warning: Changelog file not found: /github/workspace/CHANGELOG.md - creating it.
debug: update_additional_files()
Bumping with a patch version to 1.0.2
debug: set_new_version('1.0.2')
debug: Writing new version number: path=PosixPath('setup.py') pattern='__version__ *[:=] *["\\\'](\\d+\\.\\d+\\.\\d+(-beta\\.\\d+)?)["\\\']' num_matches=1
debug: set_new_version -> True
debug: commit_new_version('1.0.2')
debug: commit_new_version -> [main [64](https://github.com/shawnesquivel/pmp-otk-sandbox/actions/runs/4018206730/jobs/6903535852#step:6:65)d5a24] 1.0.2
debug:  2 files changed, 8 insertions(+), 1 deletion(-)
debug:  create mode 100644 CHANGELOG.md
debug: tag_new_version('1.0.2')
debug: tag_new_version -> 
Pushing new version
debug: get_hvcs()
debug: get_hvcs -> <class 'semantic_release.hvcs.Github'>
debug: get_hvcs()
debug: get_hvcs -> <class 'semantic_release.hvcs.Github'>
debug: push_new_version(, auth_token='***', owner='***', name='pmp-otk-sandbox', branch=main, domain='github.com')
error: Cmd('git') failed due to: exit code(128)
error:   cmdline: git push ***github.com/***/pmp-otk-sandbox.git main
error:   stderr: 'remote: Write access to repository not granted.
error: fatal: unable to access 'https://github.com/***/pmp-otk-sandbox.git/': The requested URL returned error: 403'
```
</details>

## Expected behavior

I expected to increment the version, and to not have and write access issues.

## Additional context

Access given to my token
![image](https://user-images.githubusercontent.com/94336773/214927319-78a44c04-d7a3-49ab-89a1-2e44bff1e487.png)

Example commit: 
```
git commit -m "fix: we fixed the damn bug"
```

**Comment by bernardcooke53** [comment]
Hi :wave: 

Just wanted to check - you might have a PAT with the right permissions, but it looks like the remote is denying you permission to write. Is the repo private? I'd kindly ask you to check that your branch protection rules aren't preventing you pushing to `main` and that you've got at least the `Write` repository role?

You could probably check with
```shell
git checkout main
git commit --allow-empty -m "chore: test push access"
git push -u origin main
```
if your setup allows you to test out making a commit to main like that.

Aside from this your configuration looks ok - would help if you could find out what those settings are though! Thanks

**Comment by shawnesquivel** [comment]
Hi @bernardcooke53 😀,

Yes it is a private repository. I tried out the git commands you recommended and it seemed to work for the test push. I tried it again without the --allow-empty just to see what would happen. It worked again!

<details>
<summary>test push access</summary>
```
debug: Checking commit 4aa229936c27b0bee[13](https://github.com/shawnesquivel/pmp-otk-sandbox/actions/runs/4065563921/jobs/7000441618#step:6:14)72be1c619cf9dec2f290b
debug: Checking commit af7a51af85631aa5bd7778e917ca5cebe224e244
debug: Checking commit 2d3476a19dae1759b5570fd8eb6aa2d25ab527d9
debug: Checking commit 13556bcf5bfe1ab928bf3f4a09f124f13e3638fb
debug: Checking commit f24f843e2bcc9a0862edb69d3fb1ba419a[14](https://github.com/shawnesquivel/pmp-otk-sandbox/actions/runs/4065563921/jobs/7000441618#step:6:15)[39](https://github.com/shawnesquivel/pmp-otk-sandbox/actions/runs/4065563921/jobs/7000441618#step:6:40)3b
debug: Version matches regex 1.0.2
debug: 
debug: Automatically generated by python-semantic-release
debug: get_current_release_version_by_commits -> 1.0.2
Current version: 1.0.2, Current release version: 1.0.2
debug: evaluate_version_bump('1.0.2', None)
debug: Reference v1.0.2 does not exist, considering entire history
debug: parse_commit_message('chore: test push access')
debug: parse_commit_message -> ParsedCommit(bump=0, type='chore', scope=None, descriptions=['test push access'], breaking_descriptions=[])
debug: parse_commit_message('Update PULL_REQUEST_TEMPLATE.md')
debug: Ignoring UnknownCommitMessageStyleError: Unable to parse the given commit message: Update PULL_REQUEST_TEMPLATE.md
debug: parse_commit_message('Update issue templates')
debug: Ignoring UnknownCommitMessageStyleError: Unable to parse the given commit message: Update issue templates
debug: parse_commit_message('added templates')
debug: Ignoring UnknownCommitMessageStyleError: Unable to parse the given commit message: added templates
debug: "1.0.2
debug: 
debug: Automatically generated by python-semantic-release
debug: " is commit for 1.0.2, breaking loop
debug: Commits found since last release: 4
debug: get_new_version('1.0.2', '1.0.2', None, False, True)
debug: get_new_version -> 1.0.2
debug: get_repository_owner_and_name()
debug: get_repository_owner_and_name -> ('***', 'pmp-otk-sandbox')
debug: Running publish on branch main
debug: checkout(main)
debug: checkout -> Your branch is up to date with 'origin/main'.
```
</details>

I checked my branch protection rules. They are still set to default.

![image](https://user-images.githubusercontent.com/94336773/216076872-35b9b7c1-10b4-4f65-9ed8-8ac35737e1e0.png)

I thought it was working, so I proceeded as normal and tried to make some test commits to see if it would increment the version number.

Is it possible that the issue may have been that I was missing the '-u' flag on my git pushes? Apparently not, I tried adding it, and re-adding it. I also tried merging another branch into the main branch and got other errors.

Finally, I tried the original empty commit you recommended again. From my understanding of the logs, it saw that intermediate 'bug: we fixed the bug' messages, tried to increment the version again, and couldn't do it due to the "write to repository access" issues from before.

<details>


<summary>log</summary>
```
debug: ### Fix
debug: * We fixed the bug ([`07b6f11`](https://github.com/***/pmp-otk-sandbox/commit/07b6f1142034e78ead976dfaf6330047ea259112))')
warning: Changelog file not found: /github/workspace/CHANGELOG.md - creating it.
debug: update_additional_files()
Bumping with a minor version to 1.1.0
debug: set_new_version('1.1.0')
debug: Writing new version number: path=PosixPath('setup.py') pattern='__version__ *[:=] *["\\\'](\\d+\\.\\d+\\.\\d+(-beta\\.\\d+)?)["\\\']' num_matches=1
debug: set_new_version -> True
debug: commit_new_version('1.1.0')
debug: commit_new_version -> [main 283d859] 1.1.0
debug:  2 files changed, 11 insertions(+), 1 deletion(-)
debug:  create mode 100644 CHANGELOG.md
debug: tag_new_version('1.1.0')
debug: tag_new_version -> 
Pushing new version
debug: get_hvcs()
debug: get_hvcs -> <class 'semantic_release.hvcs.Github'>
debug: get_hvcs()
debug: get_hvcs -> <class 'semantic_release.hvcs.Github'>
debug: push_new_version(, auth_token='***', owner='***', name='pmp-otk-sandbox', branch=main, domain='github.com')
error: Cmd('git') failed due to: exit code([128](https://github.com/shawnesquivel/pmp-otk-sandbox/actions/runs/4065990344/jobs/7001447792#step:6:129))
error:   cmdline: git push ***github.com/***/pmp-otk-sandbox.git main
error:   stderr: 'remote: Write access to repository not granted.
error: fatal: unable to access 'https://github.com/***/pmp-otk-sandbox.git/': The requested URL returned error: 403'
```

</details>


Here is a summary in reverse order of all the tests I did (in reverse order due to github actions)

![image](https://user-images.githubusercontent.com/94336773/216088625-009f7e73-1810-4a43-b2aa-4befcf886950.png)
